### PR TITLE
Product switch preview endpoint

### PR DESCRIPTION
**Pull request by rupertbates**
## What does this change?
This PR updates the product move endpoint to provide the ability to fetch a preview of the cost of a product switch without actually carrying it out. This will enable MMA to explain to users what the switch will cost before they commit to it.

The preview functionality is activated through a new `preview` parameter in the input json, when this is true the api will provide preview information, when it is false it will carry out the product switch.

## Example request

```shell
curl --location --request POST 'https://jlw1vyf51a.execute-api.eu-west-1.amazonaws.com/DEV/product-move/A-S00464225' \
--header 'x-api-key: replace_with_api_key' \
--header 'Content-Type: application/json' \
--data-raw '{
    "price": "10",
    "preview": true
}'
```

```json
{
    "amountPayableToday": 4.39,
    "contributionRefundAmount": -5.610000000,
    "supporterPlusPurchaseAmount": 10.000000000
}
```

### CODE REVIEW

File: handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/Subscribe.scala [hunk 1]
```diff
@@ -56,7 +56,7 @@ case class CreateSubscriptionResponse(subscriptionNumber: String)
 given JsonDecoder[CreateSubscriptionResponse] = DeriveJsonDecoder.gen[CreateSubscriptionResponse]
 
 case class ChargeOverrides(
-    price: Option[Double] = None,
+    price: Option[BigDecimal] = None,
```
Comment (rupertbates) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/Subscribe.scala:59]: I've changed any currency amounts to use `BigDecimal` for consistency with other codebases

File: handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/SubscriptionUpdate.scala [hunk 2]
```diff
@@ -50,46 +57,137 @@ private class SubscriptionUpdateLive(zuoraGet: ZuoraGet) extends SubscriptionUpd
...
+)
 case class AddRatePlan(
     contractEffectiveDate: LocalDate,
     productRatePlanId: String,
     chargeOverrides: List[ChargeOverrides],
 )
-case class RemoveRatePlan(contractEffectiveDate: LocalDate, ratePlanId: String)
-
-case class SubscriptionUpdateRequest(
+case class RemoveRatePlan(
```
Comment (rupertbates) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/SubscriptionUpdate.scala:121]: From here on down it's mostly just where I've moved things around a bit

File: handlers/product-move-api/src/test/scala/com/gu/productmove/HandlerSpec.scala [hunk 3]
```diff
@@ -52,33 +52,28 @@ import scala.language.postfixOps
 object HandlerSpec extends ZIOSpecDefault {
   def spec = {
     val time = OffsetDateTime.of(LocalDateTime.of(2022, 5, 10, 10, 2), ZoneOffset.ofHours(0)).toInstant
+    val expectedSubNameInput = "A-S00339056"
```
Comment (rupertbates) [inline handlers/product-move-api/src/test/scala/com/gu/productmove/HandlerSpec.scala:55]: All these vals were defined in each of the tests so I've moved them to object level to remove some duplication

File: handlers/product-move-api/src/test/scala/com/gu/productmove/HandlerSpec.scala [hunk 4]
```diff
@@ -90,38 +85,22 @@ object HandlerSpec extends ZIOSpecDefault {
...
+          ZLayer.succeed(new MockGetSubscription(getSubscriptionStubs())),
+          ZLayer.succeed(new MockSubscriptionUpdate(subscriptionUpdatePreviewStubs, subscriptionUpdateStubs)),
           ZLayer.succeed(new MockSQS(sqsStubs)),
           ZLayer.succeed(new MockGetAccount(getAccountStubs, getPaymentMethodStubs)),
           ZLayer.succeed(Stage.valueOf("PROD")),
         )
       },
-
-      /*
       test("productMove endpoint is successful for a refunded customer") {
```
Comment (rupertbates) [inline handlers/product-move-api/src/test/scala/com/gu/productmove/HandlerSpec.scala:95]: This test was commented out so I put it back in and made it pass

**Review by jonflynng — APPROVED** [review]
looks really good

**Comment by jonflynng** [comment]
[Preview Endpoint for Switching](https://trello.com/c/1tLPf5v1/3597-preview-endpoint-for-switching)

**Review by jonflynng — APPROVED** [review]

**Pull request merged**